#!/usr/bin/env python
PACKAGE = "bebop_linear_mpc"

#Initial MPC params for bebop quad.
#You can change this throuhg dynamic reconfigure later.
from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add("q_x",    double_t,    0, "Penality on x axis pos", 100.0 ,  1.0, 200.0)
gen.add("q_y",    double_t,    0, "Penality on y axis pos", 100.0 ,  1.0, 200.0)
gen.add("q_z",    double_t,    0, "Penality on z axis pos", 100.0 , 1.0, 200.0)

gen.add("q_vx",    double_t,    0, "Penality on x axis vel", 100.0 ,  1.0, 200.0)
gen.add("q_vy",    double_t,    0, "Penality on y axis vel", 100.0 ,  1.0, 200.0)
gen.add("q_vz",    double_t,    0, "Penality on z axis vel", 100.0 , 1.0, 200.0)

gen.add("q_roll",     double_t,    0, "Penality on roll state", 100.0 ,  1.0, 200.0)
gen.add("q_pitch",    double_t,    0, "Penality on pitch state", 100.0 , 1.0, 200.0)

gen.add("r_roll",     double_t,    0, "Penality on roll cmd", 100.0,   1.0, 200.0)
gen.add("r_pitch",    double_t,    0, "Penality on pitch cmd", 100.0 , 1.0, 200.0)
gen.add("r_thrust",   double_t,    0, "Penality on thrust cmd", 100.0 , 1.0, 200.0)

gen.add("r_droll",     double_t,    0, "Penality on delta roll cmd", 0.5 ,   0.1, 1.0)
gen.add("r_dpitch",    double_t,    0, "Penality on delta pitch cmd", 0.5 ,  0.1, 1.0)
gen.add("r_dthrust",   double_t,    0, "Penality on delta thrust cmd", 0.05 ,  0.001, 0.1)

gen.add("roll_max",     double_t,    0, "max roll cmd [rad]", 0.1, 0.01, 0.21)
gen.add("pitch_max",    double_t,    0, "max pitch cmd [rad]", 0.1, 0.01, 0.21)
gen.add("yaw_rate_max",     double_t,    0, "max yawrate cmd [rad/s]", 0.5, 0.1, 1.0)
gen.add("thrust_min",   double_t,    0, "min vz cmd [m/s]", -0.5,  -1.0, 0.0)
gen.add("thrust_max",   double_t,    0, "max vz cmd [m/s]", 0.5,  0, 1.0)

gen.add("K_yaw",   double_t,    0, "yaw gain", 5.0 ,  0.01, 10.0)
gen.add("Ki_xy",   double_t,    0, "integrator gain for xy", 2.5 ,  0.01, 5.0)
gen.add("Ki_altitude",   double_t,    0, "integrator gain for altitude", 2.5 ,  0.01, 5.0)

gen.add("enable_offset_free",   bool_t,   0, "XY offset free MPC",  False)
gen.add("enable_integrator",   bool_t,   0, "integral action",  False)

exit(gen.generate(PACKAGE, "Config", "LinearMPC"))
